File: 프로그래머스/파이썬/Lv3/등굣길.py
def solution(m, n, puddles):
    grid = [[0] * (m + 1) for _ in range(n + 1)]
    grid[1][1] = 1
    for y in range(1, n + 1):
        for x in range(1, m + 1):
            if [x, y] in puddles:
                continue
            b1 = grid[y - 1][x]
            b2 = grid[y][x - 1]
            if [x, y - 1] in puddles:
                b1 = 0
            if [x - 1, y] in puddles:
                b2 = 0
            grid[y][x] += b1 + b2
    answer = grid[n][m] % 1000000007
    return answer

File: 프로그래머스/JavaScript/Lv3/등굣길.js
function solution(m, n, puddles) {
    const DP = Array.from(Array(n + 1), () => Array(m + 1).fill(0))
    
    DP[1][1] = 1
    const check = {}
    
    for (const [x, y] of puddles) check[`${y},${x}`] = true;
    
    for (let i = 1; i < n + 1; i++){
        for (let j = 1; j < m + 1; j++){
            const v = `${i},${j}`
            if (check[v]) continue
            
            const left = check[`${i},${j - 1}`] === true ? 0 : DP[i][j - 1]
            const up = check[`${i - 1},${j}`] === true ? 0 : DP[i - 1][j]
            
            DP[i][j] += (up + left) % 1000000007
        }
    }
    

    return DP[n][m];
}